libs/datamodel/core/src/transform/ast_to_dml/db/types.rs

```diff
@@ -1,77 +1,274 @@
+    pub(super) models: HashMap<ast::ModelId, ModelData<'ast>>,
+    /// This contains only the relation fields actually present in the schema
+    /// source text.
+    pub(super) relation_fields: BTreeMap<(ast::ModelId, ast::FieldId), RelationField<'ast>>,
 }
 
-impl Types {
-    /// Detect self-referencing type aliases, possibly indirectly. We loop
-    /// through each type alias in the schema. If it references another type
-    /// alias — which may in turn reference another type alias —, we check that
-    /// it is not self-referencing. If a type alias ends up transitively
-    /// referencing itself, we create an error diagnostic.
-    pub(super) fn detect_alias_cycles(&self, ast: &ast::SchemaAst, diagnostics: &mut Diagnostics) {
-        // The IDs of the type aliases we traversed to get to the current type alias.
-        let mut path = Vec::new();
-        // We accumulate the errors here because we want to sort them at the end.
-        let mut errors: Vec<(ast::AliasId, DatamodelError)> = Vec::new();
-
-        for (alias_id, ty) in &self.type_aliases {
-            // Loop variable. This is the "tip" of the sequence of type aliases.
-            let mut current = (*alias_id, ty);
-            path.clear();
-
-            // Follow the chain of type aliases referencing other type aliases.
-            while let ScalarFieldType::Alias(next_alias_id) = current.1 {
-                path.push(current.0);
-                let next_alias = &ast[*next_alias_id];
-                // Detect a cycle where next type is also the root. In that
-                // case, we want to report an error.
-                if path.len() > 1 && &path[0] == next_alias_id {
-                    errors.push((
-                        *alias_id,
-                        DatamodelError::new_validation_error(
-                            &format!(
-                                "Recursive type definitions are not allowed. Recursive path was: {} -> {}.",
-                                path.iter().map(|id| &ast[*id].name.name).join(" -> "),
-                                &next_alias.name.name,
-                            ),
-                            next_alias.field_type.span(),
-                        ),
+#[derive(Debug)]
+enum FieldType {
+    Model(ast::ModelId),
+    Scalar(ScalarFieldType),
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum ScalarFieldType {
+    Enum(ast::EnumId),
+    BuiltInScalar(dml::scalars::ScalarType),
+    Alias(ast::AliasId),
+    Unsupported,
+}
+
+#[derive(Debug)]
+pub(crate) struct ScalarField {
+    pub(crate) r#type: ScalarFieldType,
+    pub(crate) is_ignored: bool,
+    pub(crate) is_updated_at: bool,
+    pub(crate) default: Option<dml::default_value::DefaultValue>,
+}
+
+#[derive(Debug)]
+pub(crate) struct RelationField<'ast> {
+    pub(crate) referenced_model: ast::ModelId,
+    pub(crate) on_delete: Option<dml::relation_info::ReferentialAction>,
```

Comment:
The million yen question here is should these always be set? You have the datasource in the context, so we could get the correct defaults already for both clauses here.

Reply:
At this stage I think it's good to take the same stance as for relation fields: populate only what is actually in the AST. We can have a `db.get_relation_on_delete(relation_id)` that fills in the default later.